ctrlproxy.1
.\" ctrlproxy is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH ctrlproxy 1 "October 14, 2002"
.SH NAME
ctrlproxy \- an IRC proxy with multi-server and plugin support

.SH SYNOPSIS
.B ctrlproxy
[options]
.RI
.SH DESCRIPTION
\fBctrlproxy\fP is a modular IRC proxy. It connects 
to a number of IRC servers and then passes data it receives on 
to the modules that have been loaded into it. Current 
modules are client_simple (for proxy support) and 
log (for irssi-style log file generation).

.PP
.SH OPTIONS
\fBctrlproxy\fP recognizes the following command-line options
.TP
.B -D, --daemon
Go to the background after the program has been started (daemon mode).
.TP 
.B -r, --rc-file=RCFILE
Read configuration file from specified location 
.I RCFILE
instead of from 
.I .ctrlproxyrc
in the users home directory
.TP
.B -v, --version
Show copyright and version information
.TP
.B -d, --debug
Output debug information.
.TP
.B -?, --help
Show help message with short argument information
.TP
.B --usage
Display brief usage message

.SH "PLUGINS"
.PP
Current included plugins with ctrlproxy are:

.TP
\fBclient_simple\fP
Simple client framework. Multiple clients can connect to 
a specified port and then proxying to the specified server is done. 
Has password authentication support and multiple clients can 
connect.

.TP
\fBlog\fP
Log irc traffic to a specified file, using the same style 
as the irssi(1) IRC client does.

.TP
\fBctcp\fP
CTCP replies on some CTCP messages. Currently supported:
VERSION, FINGER, CLIENTINFO, TIME and SOURCE.

.SH "SEE ALSO"
.BR irssi (1), 
.BR ctrlproxyrc (5), 
.BR http://www.nl.linux.org/~jelmer/ctrlproxy/
.SH BUGS
Probably many, but none are known at the moment. Please send 
any you find to jelmer@nl.linux.org

.SH LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
\fBWITHOUT ANY WARRANTY\fR; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU 
General Public License for more details.
.PP
You should have received a copy of the GNU General Public License 
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple PLace, Suite 330, Boston, MA  02111-1307  USA

.SH AUTHOR
.PP
 Jelmer Vernooij <jelmer@nl.linux.org>


NEWS
 Otherwise indicated differently, all changes made by 
 	Jelmer Vernooij.

Ctrlproxy 3.0.3 UNRELEASED

  BUG FIXES

    * Handle 470 response correctly.

	* Recognize supported channel / user modes correctly.
	
	* Properly store topic_set_time and topic_set_by in the 
	  linestack. (#108)

    * Reset idle time when AWAY is used in auto-away (#121)

	* Fix crash when using /CTRLPROXY without arguments

	* Don't update config when we're disconnecting during shutdown. (#113)

	* Handle 354 response correctly. (#134)

    * Handle responses after forced joins correctly. (#125)

	* Handle 250 response correctly. (#138)

	* Handle 263 response correctly. (#135)

  INTERNALS

    * Use GIOChannels for reading the motd. 
		(patch by Emanuele Giaquinta <e.giaquinta@glauco.it>)

  PORTABILITY
    
    * Include netinet/in.h in src/utils.c and src/network.c. Required 
	  for building on FreeBSD. Reported by underlig.


Ctrlproxy 3.0.2 2007-04-15

  BUG FIXES

    * Support redirecting 396, 318, 437 and 330 replies. (#84)

	* Fixed redirect warnings for 251, 252, 253, 254, 265 and 266.

    * Don't abort when trying to send auto-AWAY while network is 
	  disconnected.

	* Handle CTRLPROXY command without any arguments. (#92)

    * Fixed crash when using SSL support. (#83)

	* Add support for storing topic set time and topic set by nick (#97).

	* Fixed CTCP source reply URL. (#86)

	* Hide assertion in g_key_file_free in free_config when a network has no associated file (patch by exg)

	* Fix coredump with --init when the default config cannot be opened (patch by exg)

	* Fix crash when CHARSET command fails (#89, #96)

	* Fall back to default charset if client charset can't be set 
		(patch by Emanuele Giaquinta <e.giaquinta@glauco.it>)

	* Fix use of %S segfault in log_custom. (#111)

	* 005-lines are now forwarded to the end-client in a more 
      structural manner. (#94)
	
	* Properly expand target in log_irssi if it is a comma-separated list (#9)
        (based on patch by Korbinian Rosenegger)

    * Properly expand target in log_custom if it is a comma-separated list (#9)
        (patch by Emanuele Giaquinta and Korbinian Rosenegger)

	* Don't append extra colon before RPL_TOPICWHOTIME (Korbinian Rosenegger)

	* Adding timed backlog in BACKLOG command without a channel 
	  (Korbinian Rosenegger)

	* Handle disconnects of clients properly (#108)

	* Fix handling of non-UTF8 characters in linestack_file.

	* Don't use CHARSETs by default (#117).

	* Never send origin to the server (#122).

	* Added support for 307 responses. 

  FEATURES

    * More advanced help support

	* Add LOG_LEVEL command (#101)

	* Add 'make uninstall' (#95)

	* Do some more config checking (#102)

	* Show disconnect reason in admin network (#100)

	* Add support for handling additional 005 parameters: MAXCHANNELS,
		NICKLEN, MAXBANS, MODES, NICKLEN, MAXBANS, TOPICLEN, KICKLEN, 
		AWAYLEN, MAXTARGETS, WALLCHOPS, WALLVOICES, RFC2812, PENALTY,
		FNC, SAFELIST, USERIP, CNOTICE, KNOCK, VCHANNELS, WHOX, CALLERID,
		ACCEPT, KEYLEN, USERLEN, HOSTLEN, MAXCHANNELLEN, 
		MAXNICKLEN, DEAF, ELIST, MAXPARA, NAMESX, SECURELIST, WATCH
		REMOVE, MAP, OVERRIDE, VBANLIST and SILENCE

	* Print actual address that a listener bound to.

  INTERNALS

    * Remove is_private from line struct (#12).
		(patch by Emanuele Giaquinta <e.giaquinta@glauco.it>)

	* Implement ISON, USERHOST, WHO, WHOIS, JOIN, PART, AWAY for admin networks. (#99)

  PORTABILITY

    * Looks for check using pkg-config now.

    * Fix handling of addresses when listening on FreeBSD. (#87)

  DOCUMENTATION
  	
  	* Added more documentation to the help system. (#106)


Ctrlproxy 3.0.1 2007-01-04

  BUG FIXES

    * Logging for network state information has been simplified and been 
	  made less verbose.

	* 324 and 329 replies are now cached.

	* 352 and 315 replies are now (optionally) cached. The "max_who_age" 
	  setting can be set to the number of seconds that results should be cached.
	  This prevents "Excess Flood" errors when connecting using several xchat clients
	  at once. Can be disabled by setting max_who_age=0 (the default).

    * Fixed bug in administration command (reported by Daniel Poelzleithner)

    * Fix invalid assert in auto-away (reported by Daniel Poelzleithner)

	* Fix rejoining parted channels on restart (#78)

	* Fix erroneus toggling of "report-time" setting (#77)
	
	* Fix endless loop in logging system when debug level is 
	  set to 5 (reported by _static_).
	
	* listener configuration is no longer truncated at exit.

	* Fix channel information not being replicated for the administration 
	  network.

	* Allow ports different than 6680 in ctrlproxy --init. (#85)

  PORTABILITY
  
    * Added replacement functions for g_file_{set,get}_contents, for those 
      with GLib < 2.8
    
  DOCUMENTATION UPDATES

    * Updated command documentation (thanks, Korbinian Rosenegger)

    * Updated requirements

  MINOR CHANGES

    * Fixed several compiler warnings

Ctrlproxy 3.0	2006-12-08

  This list is not complete. Several subsystems in ctrlproxy have been 
  rewritten.

  BUG FIXES

    * NetBSD portability fixes. (Adrian Portelli)

    * Properly support strict-rfc1459 comparisons.

	* No longer depend on specific order of USER and NICK commands.

	* Fixed 100% CPU usage bug.

	* Fix large number of memory leaks.

	* Fix crash bug in log_custom (Korbinian Rosenegger)

	* Support 'true' RFC1459 string comparison.

	* Respect CASEMAPPING sent by server.

	* Support different NICK sent by client in replication.

	* Make repl_* work when NICK sent by client differs from current NICK.

	* Support mIRC and ksirc.

	* Fix bug with setting PRIVATE too often.

	* Fix issue with NickServ IDENTIFY'ing when connecting to server.
	
	* Fix problems with incorrectly added colons.

  INTERNALS

	* Reduce number of files generated by configure.

	* Moved source control system to Bazaar (http://www.bazaar-vcs.org/)


python/ctrlproxy.c
/*
	ctrlproxy: A modular IRC proxy
	(c) 2002-2007 Jelmer Vernooij <jelmer@jelmer.uk>

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <Python.h>
#include "ctrlproxy.h"

#include "libirc/python/irc.h"

extern struct global *my_global;

static PyObject *py_log_global(PyObject *self, PyObject *args)
{
	int level;
	char *text;
	if (!PyArg_ParseTuple(args, "is", &level, &text))
		return NULL;

	log_global(level, "%s", text);

	Py_RETURN_NONE;
}

static PyMethodDef ctrlproxy_methods[] = {
	{ "log_global", (PyCFunction)py_log_global, METH_VARARGS,
		"log_global(level, text)\n"
		"Log" },
	{ NULL }
};

typedef struct {
	PyObject_HEAD
} PyNetworkDictObject;

static Py_ssize_t py_network_dict_length(PyNetworkDictObject *self)
{
	return g_list_length(my_global->networks);
}

static PyObject *py_network_dict_get(PyNetworkDictObject *self, PyObject *py_name)
{
	struct irc_network *n;
	char *name;
	PyNetworkObject *ret;

	if (!PyString_Check(py_name)) {
		PyErr_SetNone(PyExc_KeyError);
		return NULL;
	}

	name = PyString_AsString(py_name);

	n = find_network(my_global->networks, name);
	if (n == NULL) {
		PyErr_SetNone(PyExc_KeyError);
		return NULL;
	}

	ret = PyObject_New(PyNetworkObject, &PyNetworkType);
	if (ret == NULL) {
		PyErr_NoMemory();
		return NULL;
	}
	Py_INCREF(self);
	ret->network = irc_network_ref(n);
	return (PyObject *)ret;
}	

static PyMappingMethods py_network_dict_mapping = {
	.mp_length = (lenfunc)py_network_dict_length,
	.mp_subscript = (binaryfunc)py_network_dict_get,
};

static PyTypeObject PyNetworkDictType = {
	.tp_name = "NetworkDict",
	.tp_basicsize = sizeof(PyNetworkDictObject),
	.tp_dealloc = (destructor)PyObject_Del,
	.tp_flags = Py_TPFLAGS_DEFAULT,
	.tp_as_mapping = &py_network_dict_mapping,
};

static PyObject *PyNetworkDict_New(void)
{
	PyNetworkDictObject *ret = PyObject_New(PyNetworkDictObject, &PyNetworkDictType);
	if (ret == NULL) {
		PyErr_NoMemory();
		return NULL;
	}
	return (PyObject *)ret;
}

void initctrlproxy(void)
{
	PyObject *m, *networkdict;

	if (PyType_Ready(&PyNetworkDictType) < 0)
		return;

	m = Py_InitModule3("ctrlproxy", ctrlproxy_methods,
					   "ControlProxy");
	if (m == NULL)
		return;

	PyModule_AddIntConstant(m, "LOG_DATA", LOG_DATA);
	PyModule_AddIntConstant(m, "LOG_TRACE", LOG_TRACE);
	PyModule_AddIntConstant(m, "LOG_INFO", LOG_INFO);
	PyModule_AddIntConstant(m, "LOG_WARNING", LOG_WARNING);
	PyModule_AddIntConstant(m, "LOG_ERROR", LOG_ERROR);
	networkdict = PyNetworkDict_New();
	if (networkdict == NULL) {
		return;
	}
	PyModule_AddObject(m, "networks", networkdict);
}
